pkg/gardenlet/controller/seed/seed/reconciler_delete.go

```diff
@@ -0,0 +1,333 @@
+	"github.com/gardener/gardener/pkg/operation/botanist/component/istio"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/kubestatemetrics"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/networkpolicies"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/nginxingress"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/resourcemanager"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/seedadmissioncontroller"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/seedsystem"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/vpa"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/vpnauthzserver"
+	seedpkg "github.com/gardener/gardener/pkg/operation/seed"
+	"github.com/gardener/gardener/pkg/utils/flow"
+)
+
+func (r *Reconciler) delete(ctx context.Context, log logr.Logger, seedObj *seedpkg.Seed) (reconcile.Result, error) {
+	seed := seedObj.GetInfo()
+
+	if !sets.NewString(seed.Finalizers...).Has(gardencorev1beta1.GardenerName) {
+		return reconcile.Result{}, nil
+	}
+
+	if seed.Spec.Backup != nil {
+		backupBucket := &gardencorev1beta1.BackupBucket{ObjectMeta: metav1.ObjectMeta{Name: string(seed.UID)}}
+		if err := r.GardenClient.Delete(ctx, backupBucket); client.IgnoreNotFound(err) != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Before deletion, it has to be ensured that no Shoots nor BackupBuckets depend on the Seed anymore.
+	// When this happens the controller will remove the finalizers from the Seed so that it can be garbage collected.
+	associatedShoots, err := controllerutils.DetermineShootsAssociatedTo(ctx, r.GardenClient, seed)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	associatedBackupBuckets, err := controllerutils.DetermineBackupBucketAssociations(ctx, r.GardenClient, seed.Name)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if len(associatedShoots) > 0 || len(associatedBackupBuckets) > 0 {
+		parentLogMessage := "Can't delete Seed, because the following objects are still referencing it:"
+
+		if len(associatedShoots) != 0 {
+			log.Info("Cannot delete Seed because the following Shoots are still referencing it", "shoots", associatedShoots)
+			r.Recorder.Event(seed, corev1.EventTypeNormal, v1beta1constants.EventResourceReferenced, fmt.Sprintf("%s Shoots=%v", parentLogMessage, associatedShoots))
+		}
+
+		if len(associatedBackupBuckets) != 0 {
+			log.Info("Cannot delete Seed because the following BackupBuckets are still referencing it", "backupBuckets", associatedBackupBuckets)
+			r.Recorder.Event(seed, corev1.EventTypeNormal, v1beta1constants.EventResourceReferenced, fmt.Sprintf("%s BackupBuckets=%v", parentLogMessage, associatedBackupBuckets))
+		}
+
+		return reconcile.Result{}, errors.New("seed still has references")
+	}
+
+	log.Info("No Shoots, ControllerInstallations or BackupBuckets are referencing the Seed, deletion accepted")
```

Comment:
As we are not checking here for referencing `ControllerInstallations` here. Maybe we can remove it from this log message to avoid confusion.
```suggestion
        log.Info("No Shoots or BackupBuckets are referencing the Seed, deletion accepted")
```